src/compas/datastructures/tree/tree.py

```diff
@@ -0,0 +1,259 @@
+
+
+class TreeNode(object):
+    """A node of a tree data structure.
+
+    Parameters
+    ----------
+    name: str, optional
+        The name of the tree ndoe.
+    attributes: dict[str, Any], optional
+        User-defined attributes of the datastructure.
+
+    Attributes
+    ----------
+    name: str
+        The name of the datastructure.
+    attributes : dict[str, Any]
+        User-defined attributes of the datastructure.
+    parent: :class:`~compas.datastructures.TreeNode`
+        The parent node of the tree node.
+    children: set[:class:`~compas.datastructures.TreeNode`]
+        The children of the tree node.
+    tree: :class:`~compas.datastructures.Tree`
+        The tree the node belongs to.
+    is_root: bool
+        True if the node is the root node of the tree.
+    is_leaf: bool
+        True if the node is a leaf node of the tree.
+    is_branch: bool
+        True if the node is a branch node of the tree.
+    acestors: generator[:class:`~compas.datastructures.TreeNode`]
+        A generator of the acestors of the tree node.
+    descendants: generator[:class:`~compas.datastructures.TreeNode`]
+        A generator of the descendants of the tree node.
+
+    """
+
+    def __init__(self, name: str = None, attributes: dict = None):
+        self.name = name
+        self.attributes = attributes or {}
+        self._parent = None
+        self._children = set()
+        self._tree = None
+
+    def __repr__(self) -> str:
+        return "<TreeNode {}>".format(self.name)
+
+    @property
+    def is_root(self):
+        return self._parent is None
+
+    @property
+    def is_leaf(self):
+        return not self._children
+
+    @property
+    def is_branch(self):
+        return not self.is_root and not self.is_leaf
+
+    @property
+    def parent(self):
+        return self._parent
+
+    @property
+    def children(self):
+        return self._children
+
+    @property
+    def tree(self):
+        return self._tree
+
+    def add(self, node: TreeNode):
+        """Add a child node to this node."""
+        if not isinstance(node, TreeNode):
+            raise TypeError("The node is not a TreeNode object.")
+        self._children.add(node)
+        node._parent = self
+        node._tree = self.tree
+        if self.tree:
+            self.tree.nodes.add(node)
+
+    def remove(self, node: TreeNode):
+        """Remove a child node from this node."""
+        self._children.remove(node)
+        node._parent = None
+        node._tree = None
+        if self.tree:
+            self.tree.nodes.remove(node)
+
+    @property
+    def acestors(self):
+        this = self
+        while this:
+            yield this
+            this = this.parent
+
+    @property
+    def descendants(self):
+        for child in self.children:
+            yield child
+            for descendant in child.descendants:
+                yield descendant
+
+    def traverse(self):
+        """Traverse the tree from this node."""
```

Comment:
Traverse can be done many ways, is it ok to have a single top level traverse method and no alternative for choice which traversal strategy?